SonarQube.TeamBuild.PreProcessor/Roslyn/EmbeddedAnalyzerInstaller.cs

```diff
@@ -0,0 +1,208 @@
+    /// However, this code is called during the pre-process stage i.e. we don't know which
+    /// version of MSBuild will be used so we have to download the analyzers even if we
+    /// can't then use them.
+    /// </remarks>
+    public class EmbeddedAnalyzerInstaller //: IAnalyzerInstaller
+    {
+        private readonly ISonarQubeServer server;
+        private readonly string localCacheDirectory;
+        private readonly ILogger logger;
+
+        public EmbeddedAnalyzerInstaller(ISonarQubeServer server, ILogger logger)
+            : this(server, GetLocalCacheDirectory(), logger)
+        {
+        }
+
+        public EmbeddedAnalyzerInstaller(ISonarQubeServer server, string localCacheDirectory, ILogger logger)
+        {
+            if (server == null)
+            {
+                throw new ArgumentNullException("server");
+            }
+            if (string.IsNullOrWhiteSpace(localCacheDirectory))
+            {
+                throw new ArgumentNullException("localCacheDirectory");
+            }
+            if (logger == null)
+            {
+                throw new ArgumentNullException("logger");
+            }
+
+            this.server = server;
+            this.localCacheDirectory = localCacheDirectory;
+            this.logger = logger;
+        }
+
+        #region IAnalyzerInstaller methods
+
+        public IEnumerable<string> InstallAssemblies(IEnumerable<Plugin> plugins)
+        {
+            if (plugins == null)
+            {
+                throw new ArgumentNullException("plugins");
+            }
+
+            if (!plugins.Any())
+            {
+                this.logger.LogInfo(RoslynResources.EAI_NoPluginsSpecified);
+                return Enumerable.Empty<string>(); // nothing to deploy
+            }
+
+            this.logger.LogInfo(RoslynResources.EAI_InstallingAnalyzers);
+
+            this.logger.LogDebug(RoslynResources.EAI_LocalAnalyzerCache, this.localCacheDirectory);
+            Directory.CreateDirectory(this.localCacheDirectory); // ensure the cache dir exists
+
+            HashSet<string> allFiles = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            foreach (Plugin plugin in plugins)
+            {
+                IEnumerable<string> files = GetPluginResourceFiles(plugin);
+                foreach(string file in files)
+                {
+                    allFiles.Add(file);
+                }
+            }
+
+            return allFiles;
+        }
+
+        #endregion
+
+        #region Private methods
+
+        /// <summary>
+        /// We want the resource cache to be in a well-known location so we can re-use files that have
+        /// already been installed (although this won't help for e.g. hosted build agents)
+        /// </summary>
+        private static string GetLocalCacheDirectory()
+        {
+            string localCache = Path.Combine(Path.GetTempPath(), ".sonarqube", ".nuget");
```

Comment:
FYI the directories names change in a later commit (to ".sq" and ".static" if I remember correctly.